src/walrus_ops.rs

```diff
@@ -0,0 +1,105 @@
+use anyhow::{anyhow, Result};
+use walrus::{
+    ir::Value, ActiveData, ActiveDataLocation, Data, DataKind, Function, FunctionBuilder,
+    FunctionKind, GlobalKind, ImportKind, ImportedFunction, InitExpr, MemoryId, Module,
+};
+
+pub(crate) fn get_active_data_segment<'a>(
+    module: &'a mut Module,
+    mem: MemoryId,
+    addr: u32,
+) -> Result<(&'a mut Data, usize)> {
+    let data = match module.data.iter().find(|&data| {
+        let DataKind::Active(active_data) = &data.kind else {
+            return false;
+        };
+        if active_data.memory != mem {
+            return false;
+        };
+        let ActiveDataLocation::Absolute(loc) = &active_data.location else {
+            return false;
+        };
+        *loc <= addr && *loc + data.value.len() as u32 > addr
+    }) {
+        Some(data) => data,
+        None => {
+            return Err(anyhow!("Unable to find data section for env ptr"));
+        }
+    };
+    let DataKind::Active(ActiveData {
+        location: ActiveDataLocation::Absolute(loc),
+        ..
+    }) = &data.kind
+    else {
+        unreachable!()
+    };
+    let data_id = data.id();
+    let offset = (addr - *loc) as usize;
+    Ok((module.data.get_mut(data_id), offset))
+}
+
+pub(crate) fn bump_stack_global<'a>(module: &'a mut Module, offset: i32) -> Result<u32> {
+    let stack_global_id = match module.globals.iter().find(|&global| {
+        if let Some(name) = &global.name {
+            name == "__stack_pointer"
+        } else {
+            false
+        }
+    }) {
+        Some(global) => global.id(),
+        None => {
+            return Err(anyhow!("Unable to find __stack_pointer global name"));
+        }
```

Comment:
FYI: `anyhow::Context` is implemented for `Option`, so rather than matching on the result of `find`, one could simply do `find(...).context("unable to find __stack_pointer global")?`.